trace_test.go

```diff
@@ -60,3 +62,56 @@ func TestBundling(t *testing.T) {
 	case <-time.After(time.Second / 5):
 	}
 }
+
+func TestNewContext_Timeout(t *testing.T) {
```

Comment:
can you add a deeper test than this?

The issue we have seen is that the code gets stuck deep inside the grpc code. The test has to ensure that the cancellations go that deep.


We have seen many go routines are stuck a conditional wait. I don't know the shallow cancellation and timeout affects this?
 
```
goroutine 86 [semacquire]: 191 times: 
sync.runtime_notifyListWait(0xc420beb7d0, 0xc400011aec)
        /usr/local/go/src/runtime/sema.go:510 +0x10b
sync.(*Cond).Wait(0xc420beb7c0)
        /usr/local/go/src/sync/cond.go:56 +0x80
istio.io/istio/vendor/google.golang.org/api/support/bundler.(*Bundler).waitUntilAllHandled(0xc420402c00, 0x100)
        /workspace/go/src/istio.io/istio/vendor/google.golang.org/api/support/bundler/bundler.go:335 +0x75
istio.io/istio/vendor/google.golang.org/api/support/bundler.(*Bundler).Flush(0xc420402c00)
        /workspace/go/src/istio.io/istio/vendor/google.golang.org/api/support/bundler/bundler.go:250 +0x8d
istio.io/istio/vendor/contrib.go.opencensus.io/exporter/stackdriver.(*traceExporter).Flush(0xc420402b40)
        /workspace/go/src/istio.io/istio/vendor/contrib.go.opencensus.io/exporter/stackdriver/trace.go:107 +0x32
istio.io/istio/vendor/contrib.go.opencensus.io/exporter/stackdriver.(*Exporter).Flush(0xc420b5da40)
        /workspace/go/src/istio.io/istio/vendor/contrib.go.opencensus.io/exporter/stackdriver/stackdriver.go:155 +0x40
istio.io/istio/mixer/adapter/stackdriver/trace.(*handler).tryFlush(0xc429b99c60)
        /workspace/go/src/istio.io/istio/mixer/adapter/stackdriver/trace/trace.go:235 +0x6a
istio.io/istio/mixer/adapter/stackdriver/trace.(*handler).HandleTraceSpan(0xc429b99c60, 0x1c8fd80, 0xc428ff4090, 0xc42e5ceaf0, 0x2, 0x2, 0x1, 0xc42e5ceaf0)
        /workspace/go/src/istio.io/istio/mixer/adapter/stackdriver/trace/trace.go:137 +0x2ba
istio.io/istio/mixer/adapter/stackdriver.(*handler).HandleTraceSpan(0xc429ebcd00, 0x1c8fd80, 0xc428ff4090, 0xc42e5ceaf0, 0x2, 0x2, 0x0, 0xc42eaabd80)
        /workspace/go/src/istio.io/istio/mixer/adapter/stackdriver/stackdriver.go:181 +0x66
istio.io/istio/mixer/template.glob..func71(0x1c8fd80, 0xc428ff4090, 0x1c65980, 0xc429ebcd00, 0xc42958af60, 0x2, 0x2, 0xbeed9a494855497b, 0x434163c35)
        /workspace/go/src/istio.io/istio/mixer/template/template.gen.go:2202 +0x143
istio.io/istio/mixer/pkg/runtime/dispatcher.(*dispatchState).invokeHandler(0xc42b4578c0, 0x0, 0x0)
        /workspace/go/src/istio.io/istio/mixer/pkg/runtime/dispatcher/dispatchstate.go:119 +0x4c7
istio.io/istio/mixer/pkg/runtime/dispatcher.(*dispatchState).(istio.io/istio/mixer/pkg/runtime/dispatcher.invokeHandler)-fm(0x0, 0x0)
        /workspace/go/src/istio.io/istio/mixer/pkg/runtime/dispatcher/session.go:231 +0x3e
istio.io/istio/mixer/pkg/pool.(*GoroutinePool).AddWorkers.func1(0xc4204b5540)
        /workspace/go/src/istio.io/istio/mixer/pkg/pool/goroutine.go:82 +0x6f
created by istio.io/istio/mixer/pkg/pool.(*GoroutinePool).AddWorkers
        /workspace/go/src/istio.io/istio/mixer/pkg/pool/goroutine.go:80 +0x6c
```

Reply:
The reason they are stuck here is that Bundler.Flush waits until all previously submitted items have been handled. Only one Handle call is allowed at a time. The timeout I added will make the Handle call return with an error (after timing out) rather than hanging forever, so after all the Handle calls have timed out, Flush will return.

Reply:
I will try to add a test that shows this.

Reply:
I understand, but the go routine actually blocked is the following, and the question is will this be unlocked by the context cancellation? 
```

goroutine 3588 [select]: 1 times: [[3588]
istio.io/istio/vendor/google.golang.org/grpc/transport.(*Stream).waitOnHeader(0xc42d615950, 0x1, 0xc42ab70840)
        /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/transport/transport.go:222 +0xff
istio.io/istio/vendor/google.golang.org/grpc/transport.(*Stream).RecvCompress(0xc42d615950, 0xc42ab70840, 0xc42a5ad998)
        /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/transport/transport.go:233 +0x2b
istio.io/istio/vendor/google.golang.org/grpc.(*csAttempt).recvMsg(0xc42ab7cc30, 0x1999720, 0xc42d0583c0, 0x0, 0x0)
        /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/stream.go:515 +0x639
istio.io/istio/vendor/google.golang.org/grpc.(*clientStream).RecvMsg(0xc429cec500, 0x1999720, 0xc42d0583c0, 0x0, 0x0)
        /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/stream.go:395 +0x43
istio.io/istio/vendor/google.golang.org/grpc.invoke(0x1c8fd80, 0xc42f5a03f0, 0x1b1f411, 0x3b, 0x1907660, 0xc42f5a02a0, 0x1999720, 0xc42d0583c0, 0xc420d2e300, 0x0, ...)
        /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/call.go:83 +0x183
istio.io/istio/vendor/google.golang.org/grpc.(*ClientConn).Invoke(0xc420d2e300, 0x1c8fd80, 0xc42f5a03f0, 0x1b1f411, 0x3b, 0x1907660, 0xc42f5a02a0, 0x1999720, 0xc42d0583c0, 0x0, ...)
        /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/call.go:37 +0x1b3
istio.io/istio/vendor/google.golang.org/grpc.Invoke(0x1c8fd80, 0xc42f5a03f0, 0x1b1f411, 0x3b, 0x1907660, 0xc42f5a02a0, 0x1999720, 0xc42d0583c0, 0xc420d2e300, 0x0, ...)
        /workspace/go/src/istio.io/istio/vendor/google.golang.org/grpc/call.go:60 +0xc1
istio.io/istio/vendor/google.golang.org/genproto/googleapis/devtools/cloudtrace/v2.(*traceServiceClient).BatchWriteSpans(0xc420d8a1e0, 0x1c8fd80, 0xc42f5a03f0, 0xc42f5a02a0, 0x0, 0x0, 0x0, 0x20, 0xc42d0583a0, 0x0)
        /workspace/go/src/istio.io/istio/vendor/google.golang.org/genproto/googleapis/devtools/cloudtrace/v2/tracing.pb.go:84 +0xd2
istio.io/istio/vendor/cloud.google.com/go/trace/apiv2.(*Client).BatchWriteSpans.func1(0x1c8fd80, 0xc42f5a03f0, 0x0, 0x0, 0x0, 0x0, 0xc421363cc0, 0x4108c8)
        /workspace/go/src/istio.io/istio/vendor/cloud.google.com/go/trace/apiv2/trace_client.go:133 +0x78
istio.io/istio/vendor/github.com/googleapis/gax-go.invoke(0x1c8fd80, 0xc42f5a03f0, 0xc42d058380, 0x0, 0x0, 0x0, 0x0, 0x1b70080, 0x4108c8, 0x20)
        /workspace/go/src/istio.io/istio/vendor/github.com/googleapis/gax-go/invoke.go:70 +0x91
istio.io/istio/vendor/github.com/googleapis/gax-go.Invoke(0x1c8fd80, 0xc42f5a03f0, 0xc42d058380, 0x0, 0x0, 0x0, 0x0, 0x0)
        /workspace/go/src/istio.io/istio/vendor/github.com/googleapis/gax-go/invoke.go:48 +0xe2
istio.io/istio/vendor/cloud.google.com/go/trace/apiv2.(*Client).BatchWriteSpans(0xc420aa1d10, 0x1c8fd80, 0xc42f5a03f0, 0xc42f5a02a0, 0x0, 0x0, 0x0, 0x1c8fd80, 0xc42f5a0390)
        /workspace/go/src/istio.io/istio/vendor/cloud.google.com/go/trace/apiv2/trace_client.go:131 +0x1a9
istio.io/istio/vendor/contrib.go.opencensus.io/exporter/stackdriver.(*traceExporter).uploadSpans(0xc420402b40, 0xc42b212588, 0x1, 0x1)
        /workspace/go/src/istio.io/istio/vendor/contrib.go.opencensus.io/exporter/stackdriver/trace.go:128 +0x48c
istio.io/istio/vendor/contrib.go.opencensus.io/exporter/stackdriver.(*traceExporter).(istio.io/istio/vendor/contrib.go.opencensus.io/exporter/stackdriver.uploadSpans)-fm(0xc42b212588, 0x1, 0x1)
        /workspace/go/src/istio.io/istio/vendor/contrib.go.opencensus.io/exporter/stackdriver/trace.go:78 +0x48
istio.io/istio/vendor/contrib.go.opencensus.io/exporter/stackdriver.newTraceExporterWithClient.func1(0x1750c00, 0xc42d23fdc0)
```